File: 2021/07/16/main.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            for j in range(i+1, len(nums)):
                if j > i + 1 and nums[j] == nums[j-1]:
                    continue

                k = j + 1
                l = len(nums) - 1
                while k < l:
                    s = nums[i] + nums[j] + nums[k] + nums[l]
                    if s == target:
                        res.append([nums[i], nums[j], nums[k], nums[l]])
                        l -= 1
                        while k < l and nums[l] == nums[l+1]:
                            l -= 1
                        k += 1
                        while k < l and nums[k] == nums[k-1]:
                            k += 1
                    elif s > target:
                        l -= 1
                    else:
                        k += 1
        return res

File: 2021/07/16/main.go
package main

import "sort"

func fourSum(nums []int, target int) [][]int {
	res := [][]int{}
	sort.Ints(nums)

	for i := 0; i < len(nums); i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		for j := i + 1; j < len(nums); j++ {
			if j > i+1 && nums[j] == nums[j-1] {
				continue
			}
			for k, l := j+1, len(nums)-1; k < l; {
				s := nums[i] + nums[j] + nums[k] + nums[l]
				if s == target {
					res = append(res, []int{nums[i], nums[j], nums[k], nums[l]})
					l--
					for k < l && nums[l] == nums[l+1] {
						l--
					}
					k++
					for k < l && nums[k] == nums[k-1] {
						k++
					}
				} else if s > target {
					l--
				} else {
					k++
				}
			}
		}
	}

	return res
}
